### PR TITLE
Connect Kube gateway part 2: command provider

### PR DESCRIPTION
**Pull request by greedy52**
Part 1:
#28312

While waiting on #28782, I've decided to break my original planned part 2 into TWO parts. 

This change includes:

- Added a `cmd.ProviderManager` to create db/kube command providers
- Make `gateway.Gateway` an interface.

### CODE REVIEW

File: lib/teleterm/gateway/interfaces.go
Comment (greedy52) [inline lib/teleterm/gateway/interfaces.go:28]: https://github.com/gravitational/teleport/pull/28312#discussion_r1254212073

Instead of breaking the Gateway interface into different types of gateway. I found it more useful to break into reader vs setter. Though open to any suggestions.
Thanks!

Comment (smallinsky) [reply lib/teleterm/gateway/interfaces.go:28]: I don't have strong preference and I'm leaving it up to you but I think that ideally we want to split Gateway per GatewayType like ` GatewayDatabase, GatewayKube` 
and return different interface implementing based on ` targetURI.IsDB() and `targetURI.IsKube()` result:
```
	case targetURI.IsDB():
            return newDatabaseGateway()

	case targetURI.IsKube():
          return newKubeGateway(...)
```		


If we know that the `RouteToDatabase()` should be called only for DB flow and KubeconfigPath() only for kube flow we can return a generic Gateway interface and in DB/Kube gateway consumer function cast/inject the specific gateway type. 


The problem with current split is that the Reader/Setter interfaces  are bit artificial in terms of "business domain" and in current design the  `RouteToDatabase` method can be called for  kube flow and  return an invalid object.

File: lib/teleterm/gateway/interfaces.go
Comment (smallinsky) [inline lib/teleterm/gateway/interfaces.go:55]: nit: godoc

File: lib/teleterm/clusters/cluster_databases.go
Comment (smallinsky) [inline lib/teleterm/clusters/cluster_databases.go:252]: The `NewDbcmdCLICmdBuilder` is a bit confusing: 
- `cmd` keyword is repeated: New...cmd...Cmd..

File: lib/teleterm/clusters/cluster_databases.go
Comment (smallinsky) [inline lib/teleterm/clusters/cluster_databases.go:250]:
```suggestion
// NewDbcmdCLICmdBuilder creates a dbcmd.MakeDbcmdCLICmdBuilder with provided
```

File: lib/teleterm/gateway/interfaces.go
Comment (greedy52) [reply lib/teleterm/gateway/interfaces.go:28]: I'll give this another shot today but no guarantee. The problem is there is no "db flow" vs "kube flow" with the current design. only "gateway" flow. (e.g `GatewayCertReissuer`, `CLICommandProvider`). Splitting to `GatewayDatabase` and `GatewayKube` means `Gateway` has to be recasted at many places.

**Comment by smallinsky** [comment]
LGTM but it looks like some places still needed to be aligned: 
```
integration/proxy/proxy_helpers.go:1: : # github.com/gravitational/teleport/integration/proxy [github.com/gravitational/teleport/integration/proxy.test]
integration/proxy/teleterm_test.go:127:11: gateway.RouteToDatabase undefined (type gateway.Gateway has no field or method RouteToDatabase)
integration/proxy/teleterm_test.go:157:11: gateway.RouteToDatabase undefined (type gateway.Gateway has no field or method RouteToDatabase) (typecheck)
/*
integration/proxy/proxy_helpers.go:1: : # github.com/gravitational/teleport/integration/proxy [github.com/gravitational/teleport/integration/proxy.test]
integration/proxy/teleterm_test.go:127:11: gateway.RouteToDatabase undefined (type gateway.Gateway has no field or method RouteToDatabase)
integration/proxy/teleterm_test.go:157:11: gateway.RouteToDatabase undefined (type gateway.Gateway has no field or method RouteToDatabase) (typecheck)
/*
integration/proxy/proxy_helpers.go:1: : # github.com/gravitational/teleport/integration/proxy [github.com/gravitational/teleport/integration/proxy.test]
integration/proxy/teleterm_test.go:127:11: gateway.RouteToDatabase undefined (type gateway.Gateway has no field or method RouteToDatabase)
integration/proxy/teleterm_test.go:157:11: gateway.RouteToDatabase undefined (type gateway.Gateway has no field or method RouteToDatabase) (typecheck)
 ```

File: lib/teleterm/gateway/interfaces.go
Comment (ravicious) [inline lib/teleterm/gateway/interfaces.go:29]: Could you explain why we need these extra interfaces in this file vs not using them and making `base`, `db` and `kube` public? Isn't the currently implementation breaking this best practice? https://github.com/golang/go/wiki/CodeReviewComments/#interfaces

File: lib/teleterm/daemon/daemon.go
Comment (ravicious) [inline lib/teleterm/daemon/daemon.go:253]: I have to say I'm not a big fan of how this provider manager looks. It seems to introduce a lot of boilerplate. Take a look at 7198abc94b – I think we can do this without introducing a new struct.

The unit tests pass with my change, but I think the integration tests are failing due to some of the changes on this branch:

```
go test ./integration -run TestTeleterm
```

```
go test ./integration/proxy -run TestALPNSNIProxyDatabaseAccess/teleterm_gateways_cert_renewal
```

File: lib/teleterm/gateway/interfaces.go
Comment (greedy52) [reply lib/teleterm/gateway/interfaces.go:29]: good question.

Here are my reasons:
- Use of gateway scatters across `api`, `clusters`, `daemon`. ideally, the interface should be used in all these places so mocks can be used in testing. If we define the interface in comsuer, we will repeat a few times.
- The first (and most problematic in my opinion) use of gateway is actually in the gateway lib, the `CLICommandProvider`:
```
// CLICommandProvider provides a CLI command for gateways which support CLI clients.                                
type CLICommandProvider interface {                                
    GetCommand(gateway Gateway) (*exec.Cmd, error)                                
}
```
- I don't think `base` should not be exposed. `db` and `kube` can be exposed as public types but need an common interface.
- provide common "casting" method

So this implementation is kinda similar to the usage of `types.Resource`. But while typing this out, not sure if this pattern justifies in this case.

@smallinsky @ravicious what do you think?

Comment (ravicious) [reply lib/teleterm/gateway/interfaces.go:29]: Ah, I see now, `gateway/interfaces.go` is needed because you cannot "upgrade" `base` to `db` or `kube`. `types.Resource` works like this only because both `Resource` and something like `AppServer` are both interfaces.

---

I'm afraid I don't know enough about Go to have an accurate opinion on this. Defining those interfaces seemed like a smell to me, hence my original comment.

[Marek asked about splitting Gateway](https://github.com/gravitational/teleport/pull/28920#discussion_r1259708226) into GatewayDatabase and GatewayKube, to which you replied that there's no "db flow" and "kube flow" (like there's a separate, say, AppServer flow and KubeServer flow), there's only "gateway flow". I don't know if this is good or not, maybe to adhere to the Go way of things it should be rearranged.

---

While looking for some examples on how to model hierarchies in Go, I found [Type embedding: Golang's fake inheritance](https://www.dolthub.com/blog/2023-02-22-golangs-fake-inheritance/):

> Last week [a complaint from a gopher about his coworkers bringing Java conventions to a Golang codebase](https://www.reddit.com/r/golang/comments/112vfr3/how_to_deal_with_java_developers_polluting_the_go/) made a minor splash on Reddit. Their complaints:
> > (…) All constructor functions return exported interfaces which actually return unexported concrete types. I'm told this is done for encapsulation purposes, otherwise users will not be forced to use the constructor functions. So there is only ever one implementation of an interface, it is implemented by the lower case struct named the same as the exported interface. I really don't like this pattern.

Isn't this what we do here with `Gateway` and `base`?

Ultimately I'd ask someone more knowledgeable in Go about an opinion.

**Review by ravicious — APPROVED** [review]
This looks good to me, but as I mentioned in my comment, I'm not able to tell if this way of modelling a hierarchy is going to cause problems in the future or not. I'd try to seek another opinion.

**Review by smallinsky — APPROVED** [review]

File: lib/teleterm/gateway/interfaces.go
Comment (greedy52) [reply lib/teleterm/gateway/interfaces.go:29]: I will merge this for now to unblock other changes. But I will keep an on this. I don't think the problem we are solving here is tied to this gateway struct/interface but more about how teleterm lib flows are structured. I will give it more thought and maybe ask around too.

**Pull request merged**
